experimental/iterators/lib/Conversion/IteratorsToLLVM/IteratorsToLLVM.cpp

```diff
@@ -196,6 +229,530 @@ struct PrintOpLowering : public ConversionPattern {
+  StringRef funcName = funcNameAttr.getLeafReference();
+  ModuleOp module = originalOp->getParentOfType<ModuleOp>();
+  assert(module);
+  MLIRContext *context = rewriter.getContext();
+
+  // Create function op.
+  OpBuilder::InsertionGuard guard(rewriter);
+  rewriter.setInsertionPointToStart(module.getBody());
+
+  FunctionType funcType = FunctionType::get(context, inputTypes, returnTypes);
+  FuncOp funcOp =
+      rewriter.create<FuncOp>(originalOp->getLoc(), funcName, funcType);
+  funcOp.setPrivate();
+
+  // Create initial block.
+  SmallVector<Location, 4> locs(inputTypes.size(), originalOp->getLoc());
+  rewriter.createBlock(&funcOp.getBody(), funcOp.begin(), inputTypes, locs);
+
+  return funcOp;
+}
+
+/// Creates an Open function for originalOp given the provided opInfo.
+FuncOp createOpenFunc(Operation *originalOp, RewriterBase &rewriter,
+                      const IteratorAnalysis::IteratorInfo &opInfo) {
+  return createOpenNextCloseFunc(originalOp, rewriter, opInfo.stateType,
+                                 opInfo.stateType, opInfo.openFunc);
+}
+
+/// Creates an Next function for originalOp given the provided opInfo.
+FuncOp createNextFunc(Operation *originalOp, RewriterBase &rewriter,
+                      const IteratorAnalysis::IteratorInfo &opInfo,
+                      Type elementType) {
+  return createOpenNextCloseFunc(
+      originalOp, rewriter, opInfo.stateType,
+      {opInfo.stateType, rewriter.getI1Type(), elementType}, opInfo.nextFunc);
+}
+
+/// Creates an Close function for originalOp given the provided opInfo.
+FuncOp createCloseFunc(Operation *originalOp, RewriterBase &rewriter,
+                       const IteratorAnalysis::IteratorInfo &opInfo) {
+  return createOpenNextCloseFunc(originalOp, rewriter, opInfo.stateType,
+                                 opInfo.stateType, opInfo.closeFunc);
+}
+
+//===----------------------------------------------------------------------===//
+// Conversion of iterator ops to LLVM
+//===----------------------------------------------------------------------===//
+
+/// Converts the given SampleInputOp to LLVM using the converted operands from
+/// the upstream iterator.
+FailureOr<Optional<Value>>
+convertIteratorOp(SampleInputOp op, ArrayRef<Value> operands,
+                  RewriterBase &rewriter,
+                  const IteratorAnalysis &typeAnalysis) {
+  assert(op->getNumResults() == 1);
+  StreamType streamType = op->getResult(0).getType().dyn_cast<StreamType>();
+  assert(streamType);
+  Type elementType = streamType.getElementType();
+
+  auto maybeOpInfo = typeAnalysis.getIteratorInfo(op);
+  assert(maybeOpInfo.hasValue());
+  auto opInfo = maybeOpInfo.getValue();
+
+  // Create converted state.
+  Value convertedState =
+      rewriter.create<UndefOp>(op->getLoc(), opInfo.stateType);
+
+  // Create Open function. --------------------------------------------------
+  {
+    FuncOp funcOp = createOpenFunc(op, rewriter, opInfo);
```

Comment:
It's nice that the function declaration is separate, I would add another level of helpers where the creation of the body of the function is also separate so that the convertIteratorOp only has:
```
createOpenFunc(...);
buildOpenFuncBody(SampleInputOp op);
createNext
etc.
```
In the current form, it is hard to distinguish simple boilerplate from more load-bearing logic.

Reply:
Each of the "buildxxxBody" should also have some pseudo-IR snippet comment to show what is produced

Reply:
OK, done in quite a big refactor among the latest commits. This was a good investment! Check out whether you like it :)